accepted/future-releases/language-versioning/package-config-file-v2.md

```diff
@@ -0,0 +1,193 @@
+  - `packageUri`: Optional. A URI reference consisting of just a relative URI path. It should be resolved against the root directory to specify the *package URI directory*, which must be a sub-directory of the root directory. If the result of this resolution does not end with a `/`, a `/` is appended. This is the directory containing files available to Dart programs using `package:packageName/...` URIs. If omitted, the root directory is also the package URI directory. The value is a URI path, and tools must support URI references with `%`-escapes and `..` segments properly, and validate that the result of resolving it against the root directory stays inside the root directory (which a leading `..` can violate).
+  - `languageVersion`: Optional. A string containing a Dart language version consisting of a decimal numeral, a `.` character and another decimal numeral, where a decimal numeral must not have a leading `0` unless the numeral is exactly `0`. That is, a string matched in its entirety by the regular expression `(0|[1-9]\d*)\.(0|[1-9]\d*)`. *This is the same format accepted by `// @dart = x.y` comments in source files.*
+
+It is an error if any file or directory belongs to more than one package. This situation happens if two packages have specified root directories where one is a subdirectory of the other, or they are even the same directory. If that happens, the configuration is invalid and must be rejected.
+
+The following optional properties of the top-level object allows the generator to specify metadata about the file. They can all be omitted, but if they are present, they should have the expected format.
+
+- `generated`: A string containing an ISO-8601 UTC timestamp for when the file was generated, using the format: `YYYY-MM-DDTHH:mm:ss.sssZ`, for example `"generated": "2019-09-24T12:34:56.000Z"`.
+- `generator`:  A string with a name/identifier of the generator which created the file, typically `"pub"`.
+- `generatorVersion`: A string with the version of the generator, if the generator wants to remember that information. The version must be a [Semantic Version](https://semver.org/spec/v2.0.0.html). Pub can use the SDK version. Example: `"generatorVersion": "2.5.0"`.
+
+The JSON properties may occur in any order, except that the `apiVersion` *must* be the first entry in the file. The Dart JSON serializer uses iteration order of the incoming map, so controlling the order is possible. The file *should* be "pretty printed" with newlines and indentation, to make human inspection easier, but tools consuming the file must accept JSON with or without any optional whitespace.
```

Comment:
What does this buy us?

Reply:
Which "this"?

Putting the `apiVersion` first allows parsers to be more efficient than parsing the entire file first and then reading the intermediate maps. If the `apiVersion` changes, then you can know immediately which following format to expect, and use a faster parser specialized for that format ... if you want to. I like to have the option.

As for pretty-printing, it's just easier to read for humans.

Reply:
Sorry, I meant the apiVersion.  Requiring this means that you can't just use an off the shelf json serializer, right?  I guess we don't expect there to be too many generators, but still.  Maybe if we want this first, then the outermost json object should be an array?

Reply:
Changed to a recommendation instead of a requirement.